org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java

```diff
@@ -279,110 +274,128 @@ private static boolean capabilitiesComply(LanguageServerWrapper wrapper,
 	 *
 	 * @param file
 	 * @param request
-	 * @return
+	 * @return the matching LS wrappers
 	 * @throws IOException
 	 * @noreference This method is currently internal and should only be referenced
 	 *              for testing
 	 */
 	@NonNull
-	public static Collection<LanguageServerWrapper> getLSWrappers(@NonNull IFile file,
-			@Nullable Predicate<ServerCapabilities> request) throws IOException {
-		LinkedHashSet<LanguageServerWrapper> res = new LinkedHashSet<>();
-		IProject project = file.getProject();
+	public static Collection<LanguageServerWrapper> getLSWrappers(@NonNull final IFile file,
+			@Nullable final Predicate<ServerCapabilities> request) throws IOException {
+		final var project = file.getProject();
 		if (project == null) {
-			return res;
+			return Collections.emptyList();
 		}
 
-		res.addAll(getMatchingStartedWrappers(file, request));
+		final var res = new LinkedHashSet<LanguageServerWrapper>(getMatchingStartedWrappers(file, request));
 
 		// look for running language servers via content-type
-		Queue<IContentType> contentTypes = new LinkedList<>();
-		Set<IContentType> addedContentTypes = new HashSet<>();
-		contentTypes.addAll(LSPEclipseUtils.getFileContentTypes(file));
-		addedContentTypes.addAll(contentTypes);
+		final var contentTypes = new ArrayDeque<IContentType>(LSPEclipseUtils.getFileContentTypes(file));
+		final var processedContentTypes = new HashSet<IContentType>(contentTypes.size());
+		final var lsRegistry = LanguageServersRegistry.getInstance();
 
 		while (!contentTypes.isEmpty()) {
-			IContentType contentType = contentTypes.poll();
-			if (contentType == null) {
+			final var contentType = contentTypes.poll();
+			if (contentType == null || processedContentTypes.contains(contentType)) {
 				continue;
 			}
-			for (ContentTypeToLanguageServerDefinition mapping : LanguageServersRegistry.getInstance().findProviderFor(contentType)) {
-				if (mapping != null && mapping.getValue() != null && mapping.isEnabled()) {
-					LanguageServerWrapper wrapper = getLSWrapper(project, mapping.getValue(), file.getFullPath());
-					if (capabilitiesComply(wrapper, request)) {
-						res.add(wrapper);
-					}
+
+			for (final ContentTypeToLanguageServerDefinition mapping : lsRegistry.findProviderFor(contentType)) {
+				if (mapping == null || !mapping.isEnabled()) {
+					continue;
+				}
+				final LanguageServerDefinition serverDefinition = mapping.getValue();
+				if (serverDefinition == null) {
+					continue;
+				}
```

Comment:
this code is repeated below, maybe extract it to a method 
```
isValidMapping(ContentTypeToLanguageServerDefinition mapping) {
                                if (mapping == null || !mapping.isEnabled()) {
					return false;
				}
				final LanguageServerDefinition serverDefinition = mapping.getValue();
				if (serverDefinition == null) {
					return false;
				}
}
```
or similar?

Reply:
I gave it a try it realized that the serverDefinition variable is used by the subsequent code and thus there isn't much saved.

Reply:
ACK